[AMDGPU] Move WWM register pre-allocation to during regalloc

perlfu opened the pull request:
Move SIPreAllocateWWMRegs pass to just before VGPR allocation. This saves recomputation of the virtual matrix and live reg map, with the slight regression in O0 that live intervals and slot indexes must be computed.

perlfu committed:
[AMDGPU] Move WWM register pre-allocation to during regalloc

Move SIPreAllocateWWMRegs pass to just before VGPR allocation.
This saves recomputation of the virtual matrix and live reg map,
with the slight regression in O0 that live intervals and slot
indexes must be computed.

cdevadas commented on llvm/test/CodeGen/AMDGPU/global_atomics_scan_fadd.ll:
```diff
@@ -725,36 +725,36 @@ define amdgpu_kernel void @global_atomic_fadd_uni_address_div_value_agent_scope_
-; GFX9-DPP-NEXT:    v_add_f32_e32 v3, v3, v5
-; GFX9-DPP-NEXT:    v_bfrev_b32_e32 v5, 1
+; GFX9-DPP-NEXT:    v_mov_b32_dpp v42, v40 row_shr:1 row_mask:0xf bank_mask:0xf
+; GFX9-DPP-NEXT:    v_add_f32_e32 v40, v40, v42
```
It looks like the allocation is bad now, consuming more VGPRs. Earlier v31 was the higher VGPR used, now that has become v42. The same is true for some other tests. Have you checked if it is compromising the occupancy?

perlfu replied:
I have addressed this by turning off UsedPhysRegMask testing in isPhysRegUsed.
Because the pass now runs after the first register rewrite (for SGPRs) the clobbers from calls have been inserted into the UsedPhysRegMask even for VGPRs.
Ignoring seems safe as: 1) we never knew about them before; 2) they will be correctly spilt if live ranges are split.
Additionally as far as I understand the only thing in our code path that updates UsedPhysRegMask is VirtRegMap, so ignoring the map for WWM register assignment is valid.

arsenm reviewed (approved):
Really the allocator should be handling this, so as we move towards that point it makes sense to move where this runs

perlfu committed:
Skip UsedPhysRegMask test when calling isPhysRegUsed.
We are not concerned about clobbers from calls as these will be
spilt correctly -- and these were never considered before.

perlfu committed:
Ignore AMDGPU::SI_SPILL_S32_TO_VGPR in STRICT_WWM regions.

perlfu commented:
I realized that we need to explicitly ignore SI_SPILL_S32_TO_VGPR instructions which occur in STRICT_WWM sections otherwise they will unintentionally be assigned WWM registers.
This brings the total test diff on this down massive to primarily wwm-reserved-spill.ll, which shows a change in use of registers because SGPR spill now get first access to VGPRs before WWM pre-allocation runs.

This should now be good to go.

perlfu committed:
Add comment for SkipRegMaskTest.

perlfu commented on llvm/lib/Target/AMDGPU/SIPreAllocateWWMRegs.cpp:
```diff
@@ -101,7 +99,7 @@ bool SIPreAllocateWWMRegs::processDef(MachineOperand &MO) {
 
   for (MCRegister PhysReg : RegClassInfo.getOrder(MRI->getRegClass(Reg))) {
-    if (!MRI->isPhysRegUsed(PhysReg) &&
+    if (!MRI->isPhysRegUsed(PhysReg, true) &&
```
Fixed, thanks.

perlfu commented:
If there is no further feedback I will merge this on November 6th.

jayfoad commented:
Change looks good overall but I don't quite understand this:

> I realized that we need to explicitly ignore SI_SPILL_S32_TO_VGPR instructions which occur in STRICT_WWM sections otherwise they will unintentionally be assigned WWM registers. This brings the total test diff on this down massive to primarily wwm-reserved-spill.ll, which shows a change in use of registers because SGPR spill now get first access to VGPRs before WWM pre-allocation runs.

Do SI_SPILL_S32_TO_VGPR instructions refer to _physical_ VGPRs? If so, why do you need to explicitly ignore them in this pass?

If SI_SPILL_S32_TO_VGPR instructions refer to virtual VGPRs then I don't understand how they "get first access to [physical?] VGPRs".

perlfu commented:
> Change looks good overall but I don't quite understand this:
> 
> > I realized that we need to explicitly ignore SI_SPILL_S32_TO_VGPR instructions which occur in STRICT_WWM sections otherwise they will unintentionally be assigned WWM registers. This brings the total test diff on this down massive to primarily wwm-reserved-spill.ll, which shows a change in use of registers because SGPR spill now get first access to VGPRs before WWM pre-allocation runs.
> 
> Do SI_SPILL_S32_TO_VGPR instructions refer to _physical_ VGPRs? If so, why do you need to explicitly ignore them in this pass?
> 
> If SI_SPILL_S32_TO_VGPR instructions refer to virtual VGPRs then I don't understand how they "get first access to [physical?] VGPRs".

SI_SPILL_S32_TO_VGPR can take a physical or virtual VGPR.
The decision to use a physical register is made during SGPR spill lowering based on whether the function is an entry function or not.
In the entry function, physical VGPRs will be assigned for VGPR spill registers.

So a high level we currently have two independent code paths which can allocate physical registers for WWM use.
Unifying them is beyond the scope of this change, but is a near term objective which could be built on this patch.

cdevadas commented:
> The decision to use a physical register is made during SGPR spill lowering based on whether the function is an entry function or not.
One correction. I always wanted the SGPR spill lowering to use virtual VGPR here so that RA can allocate them efficiently. Currently, we use physical registers for CSR SGPR spills inserted at the prolog/epilog and virtual for the rest of the spills (the one introduced during SGPR-regalloc). This is to ensure that the CSR spills get VGPRs that won't be reused by RA for any other liveranges in the function. This is a fixup to ensure the CSR spills get static CFI entries. We earlier encountered an issue in the rocgdb during frame unwinding that the CFI entries for the CSR spills are broken.

Skipping AMDGPU::SI_SPILL_S32_TO_VGPR & AMDGPU::SI_RESTORE_S32_FROM_VGPR here is the right thing. Because they are meant to be allocated during RA.
BTW, why SI_RESTORE_S32_FROM_VGPR isn't included here?

perlfu commented:
> Skipping AMDGPU::SI_SPILL_S32_TO_VGPR & AMDGPU::SI_RESTORE_S32_FROM_VGPR here is the right thing. Because they are meant to be allocated during RA. BTW, why SI_RESTORE_S32_FROM_VGPR isn't included here?

SI_RESTORE_S32_FROM_VGPR is not included here because it does not define any VGPRs, so WWM pre-allocate will ignore it as-is.

cdevadas commented:
> > Skipping AMDGPU::SI_SPILL_S32_TO_VGPR & AMDGPU::SI_RESTORE_S32_FROM_VGPR here is the right thing. Because they are meant to be allocated during RA. BTW, why SI_RESTORE_S32_FROM_VGPR isn't included here?
> 
> SI_RESTORE_S32_FROM_VGPR is not included here because it does not define any VGPRs, so WWM pre-allocate will ignore it as-is.

Ok.
